feat: Add support for controller skeletons in Kotlin

belu opened the pull request:
Adds support to create a skeleton for Kotlin.

A major refactoring of the SkeletonBuffer was necessary:
- split up into "construct" phase (SkeletonBuffer, which constructs a context for SkeletonCreator) and "create" phase (SkeletonCreator)
- create phase has no side-effects (everything is now determined from the given input)
- Kotlin and Java are two different create-phases
- added tests that validate the whole skeleton (until now only the "package" line was validated for Java)
- more tests verifying the whole skeleton could be added easily (e.g. for Java, too)

### Issue

Fixes #259 

### Progress

- [x] Change must not contain extraneous whitespace
- [x] License header year is updated, if required
- [x] Verify the contributor has signed [Gluon Individual Contributor License Agreement (CLA)](https://docs.google.com/forms/d/16aoFTmzs8lZTfiyrEm8YgMqMYaGQl0J8wA0VJE2LCCY)

belu commented:
Screenshot of the proposed functionality:
![scenebuilder_skeleton_kotlin](https://user-images.githubusercontent.com/1174164/110250148-22b5a780-7f7a-11eb-9573-7a9f46210399.png)

abhinayagarwal commented on kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java:
```diff
@@ -122,13 +124,18 @@ public void openWindow() {
+        languageChoiceBox.getSelectionModel().select(SkeletonSettings.LANGUAGE.JAVA);
 
-        formatCheckBox.selectedProperty().addListener((ChangeListener<Boolean>) (ov, t, t1) -> update());
+        ChangeListener<Object> updateOnChangeListener = (ov, t, t1) -> update();
```
`updateOnChangeListener` could be replaced by `fxomDocumentRevisionListener`

belu replied:
Fixed. Had to change the generic on `fxomDocumentRevisionListener` from `Number` to `Object`.

abhinayagarwal commented on kit/src/test/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonBufferTest.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016, 2017 Gluon and/or its affiliates.
+ * Copyright (c) 2016, 2017, 2021 Gluon and/or its affiliates.
```
This should be `2016, 2021, Gluon...`

belu replied:
Fixed.

abhinayagarwal commented on kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonContext.java:
```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2016, 2021, Gluon and/or its affiliates.
+ * Copyright (c) 2012, 2014, Oracle and/or its affiliates.
```
New files need to have only the year when they were created. Therefore the above 2 lines can be replaced with:

```
Copyright (c) 2021, Gluon and/or its affiliates.
```

belu replied:
Fixed.

abhinayagarwal commented:
Overall the PR looks good to me. Great work :+1:

belu commented:
Hi @abhinayagarwal, thanks for the first review! Fixed the minor suggestions and am now reducing the duplication by moving  code into a base class.

To accomplish this I want to add more tests to assure that the refactoring does not break things.
However, if I use a more complex FXML file (e.g. with a `Button`) then loading fails with the following error:
`java.lang.IllegalStateException: Toolkit not initialized`

In another test I saw that you do the following:
```
@BeforeClass
public static void initialize() {
    new JFXPanel();
}
```

Do you use this pattern to initialize the toolkit? According to a first test, this seems to work.

belu commented on kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonCreatorJava.java:
```diff
@@ -0,0 +1,220 @@
+import java.lang.reflect.TypeVariable;
+import java.util.Map;
+
+public class SkeletonCreatorJava {
```
Fixed. Could you please review. What is left are some TODOs in `SkeletonCreatorKotlin`. Will try to finish this in the next days, including some tests. Thanks!

abhinayagarwal commented:
@belu Its most probably better to skip using Swing dependent toolkit initialization and use pure JavaFX instead. You can find one of the ways to do it in SceneBuilder's [StaticLoadTest](https://github.com/gluonhq/scenebuilder/blob/master/kit/src/test/java/com/oracle/javafx/scenebuilder/kit/fxom/StaticLoadTest.java#L58).

belu commented:
Hi @abhinayagarwal, thanks for the hint.

As this is done in several classes I refactored this logic into a separate helper class `JfxInitializer`. This makes it very easy for new test classes to initialize the JavaFX toolkit if necessary.

There was also one test class which used the trick with the `JFXPanel`. I also changed that class to use the `JfxInitializer`.

belu commented:
Hey @abhinayagarwal, just commited the last thing missing: support and tests for nested classes in Kotlin.
I also added a test case for nested classes in Kotlin and Java.

From my side the feature request is now complete. Could you please review and provide feedback? Thanks.

belu commented:
Hi @abhinayagarwal, when are you able to review?

abhinayagarwal commented on kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonCreator.java:
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2016, 2021, Gluon and/or its affiliates.
+ * Copyright (c) 2012, 2014, Oracle and/or its affiliates.
```
New files doesn't need entry for Oracle. Check [AbstractSkeletonCreator.java](https://github.com/gluonhq/scenebuilder/pull/328/files#diff-0d03d25c4354d83495020d140ffb31145769d5538a5a3a40420351376a4f4356R2) for correct license header.

abhinayagarwal commented on kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonCreatorJava.java:
```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2016, 2021, Gluon and/or its affiliates.
+ * Copyright (c) 2012, 2014, Oracle and/or its affiliates.
```
New files doesn't need entry for Oracle. Check [AbstractSkeletonCreator.java](https://github.com/gluonhq/scenebuilder/pull/328/files#diff-0d03d25c4354d83495020d140ffb31145769d5538a5a3a40420351376a4f4356R2) for correct license header.

abhinayagarwal commented on kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonCreatorKotlin.java:
```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2016, 2021, Gluon and/or its affiliates.
+ * Copyright (c) 2012, 2014, Oracle and/or its affiliates.
```
New files doesn't need entry for Oracle. Check [AbstractSkeletonCreator.java](https://github.com/gluonhq/scenebuilder/pull/328/files#diff-0d03d25c4354d83495020d140ffb31145769d5538a5a3a40420351376a4f4356R2) for correct license header.

abhinayagarwal commented on kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonSettings.java:
```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2016, 2021, Gluon and/or its affiliates.
+ * Copyright (c) 2012, 2014, Oracle and/or its affiliates.
```
New files doesn't need entry for Oracle. Check [AbstractSkeletonCreator.java](https://github.com/gluonhq/scenebuilder/pull/328/files#diff-0d03d25c4354d83495020d140ffb31145769d5538a5a3a40420351376a4f4356R2) for correct license header.

abhinayagarwal commented on kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java:
```diff
@@ -1,4 +1,5 @@
 /*
+ * Copyright (c) 2021, Gluon and/or its affiliates.
```
This should be "Copyright (c) 2016, 2021, Gluon and/or its affiliates."

abhinayagarwal commented on kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonWindowController.java:
```diff
@@ -122,22 +124,26 @@ public void openWindow() {
-    private final ChangeListener<Number> fxomDocumentRevisionListener
-            = (observable, oldValue, newValue) -> update();
+    private final ChangeListener<Object> fxomDocumentRevisionListener
+        = (observable, oldValue, newValue) -> update();
```
An `InvalidationListener` might be a better fit here.

belu commented:
Hi @abhinayagarwal, all your findings are fixed. Should I squash all the commits or are you doing that?

abhinayagarwal commented on kit/src/main/java/com/oracle/javafx/scenebuilder/kit/skeleton/SkeletonCreator.java:
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2016, 2021, Gluon and/or its affiliates.
```
New files only need the current year as they didn't exist back in 2016 ;)

> Copyright (c) 2021, Gluon and/or its affiliates.

Please do the same for all new files.

belu replied:
Fixed ... hopefully finally. It would be much easier without the copyright headers ;-)

abhinayagarwal commented:
LGTM! 

We will squash and merge the PR.

jperedadnr reviewed (approved):
Looks good to me.
